scripts/tools/silabs/FactoryDataProvider.py

```diff
@@ -0,0 +1,212 @@
+    IterationCount_ui = args.spake2_iteration
+    Salt_str = args.spake2_salt
+
+    if args.spake2_verifier:
+        Verifier_str = args.spake2_verifier
+    elif args.gen_spake2p_path:
+        Verifier_str = generate_spake2p_verifier(args)
+
+    if args.mcu_family:
+        mcuFamily = args.mcu_family
+
+    if args.jtag_serial:
+        jtagSerial = args.jtag_serial
+
+
+def add_SerialNo_To_CMD(cmdList):
+    """ Add the jtag serial command to the commander command
+
+        Args:
+        The commander command in list format
+    """
+    if jtagSerial:
+        cmdList.extend(["--serialno", jtagSerial])
+
+
+def create_nvm3injected_image():
+    """ Use commander command lines create a binary flashable to the EFR32 
+        containing the factory commissioning data in NVM3 section
+    """
+    global isDeviceConnected
+
+    # Retrieve the device current nvm3 data in a binary file
+    # It will be used as base to add the new credentials
+    inputImage = TEMP_FILE
+    cmd = ['commander', 'nvm3', 'read', '-o', inputImage, ]
+    add_SerialNo_To_CMD(cmd)
+    results = subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+    if results.returncode != 0:
+        # No nvm3 section found. Retrieve the device info
+        cmd = ['commander', 'device', 'info', ]
+        add_SerialNo_To_CMD(cmd)
+        try:
+            output = subprocess.check_output(cmd)
+            output = output.decode('utf-8').splitlines()
+            deviceInfo = dict(map(str.strip, lines.split(':')) for lines in output[0:len(output)-1])
+            # Only MG12 and MG24 are supported in matter currently
+            if "EFR32MG12" in deviceInfo["Part Number"]:
+                inputImage = "base_matter_mg12_nvm3.s37"
+            elif "EFR32MG24" in deviceInfo["Part Number"]:
+                inputImage = "base_matter_mg24_nvm3.s37"
+            else:
+                raise Exception('Invalid MCU')
+        except:
+            isDeviceConnected = False
+            print("Device not connected")
+            # When no device is connected user needs to provide the mcu family for which those credentials are to be created
+            if mcuFamily:
+                if "EFR32MG12" in deviceInfo["Part Number"]:
+                    inputImage = "matter_mg12_nvm3.s37"
+                elif "EFR32MG24" in deviceInfo["Part Number"]:
+                    inputImage = "matter_mg24_nvm3.s37"
+            else:
+                print("Connect debug port or provide the mcu_family")
+                return
+
+    # Convert interger to little endian hex format and strings to hex byte array format for nvm3 storage
+    Passcode = Passcode_ui.to_bytes((Passcode_ui.bit_length() + 7) // 8, "little").hex()
+    Spake2pIterationCount = IterationCount_ui.to_bytes((IterationCount_ui.bit_length() + 7) // 8, 'little').hex()
+    Discriminator = Discriminator_ui.to_bytes((Discriminator_ui.bit_length() + 7) // 8, 'little').hex()
```

Comment:
The bit length here may be confusing/wrong/not portable. The underlying types are fixed width. Suggest using fixed width explicitly